Contact information and contact us page

JessSumner opened the pull request:
A contact information template has been added to the header of each page.
This has been formatted to match the logo so the header is balanced.

The contact us page contains a simple form which saves data to the database.
A mailer is set up to send the message to the admin of the rescue.

JessSumner committed:
February flyer added to the assets file

February flyer loads up correctly in the view. No changes to controller needed.

JessSumner committed:
Set up infrastructure for Contact Us page

Generated a model.
Database table created with columns for name, email, phone and message.
Validations for these columns as all fields should be filled out. Validations for format for email and phone.
Generated a controller.

JessSumner committed:
database migrated

JessSumner committed:
Updated the column phone to cell

This matches with the other models in this application. Also a validation method has been created for cell so this can be utilized here.

JessSumner committed:
Basic content added to contact page

form_for has been used so a visitor can send a message to the rescue. Form fields include name, email, cell and message
Added a rat picture to the page

JessSumner committed:
Send button and formatting

Send button has been added to the bottom of the form.
Page has been styled to have a title, and the form and rat picture side by side.

JessSumner committed:
Added create method and errors

Create method to show flash message if the message gets saved. If the message is incorrectly filled out a flash message will appear and the visitor will be directed back to their form.

JessSumner committed:
Added a validation for the message.

The message can't be blank.

On checking that the validation was working correctly I noticed that the error styling was not applying to it because the input is a textarea instead of textfield.

The errors.scss file has been updated to include textarea as a child of field-with-errors. It correctly displays if not message is written.

JessSumner committed:
Changed Contact to ContactMessage

Whilst writing this section I started to feel that ContactMessage better described what this model and controller were doing.

All files have been changed and the site is working correctly locally.

JessSumner committed:
Mailer infrastructure created.

The mailer, html and text views and preview has been set up to handle sending the recieved contact_message to an Any Rat Rescue administrator.

JessSumner committed:
Content for html and text emails added.

Followed the structure of the emails for the adoption and volunteer applications.
Created a form fields constant.
The constant is an array and is iterated over to provide the table contents within the mailer view.

JessSumner committed:
Contact details added to the header.

Created a similar styled div to match the logo.
Contains a ul of contact information.
li have been divided into types and details and these have been formatted differently.

JessSumner committed:
Replaced the border property with box-shadow

For the logo and the information div in the header. I have put mirrored the shadow.

JessSumner committed:
Added method to send email on save of form.

Method added the contact_message_controller file.

JessSumner committed:
Removed color literals and whitespace

Migration file had trailing whitespace. This has been removed.
The box shadow was using a rgb color literal. This has been extracted into a variable. $black is now defined and $shadow is defined as 75% transparent black.

JessSumner commented:
@MattMSumner @halogenandtoast @kenyonj. Here is another PR, most of the content is principles which you have commented on already so I hope it is good to go. 
I am aware of a bug which I am troubleshooting. After submitting the contact message the visitor gets redirected to the home page. When the home page loads the text is not justified. On a page refresh it loads correctly. I have not seen this happening when I hop from page to page just on the redirect. From issues I have had before I am thinking it might be a Turbolinks problem. I am going to continue looking into it and see if I can find the problem.
On further inspection, this bug is not isolated to the Contact message form. It is also happening after the redirect from the Volunteer application but the alignment was less distorted so I only noticed now that I am looking for it.
Thank you very much for your help and guidance, it is much appreciated. :)

halogenandtoast commented:
@JessSumner my guess is that your flash messages are somehow breaking your alignment.

halogenandtoast commented on app/views/new_contact_message_mailer/contact_message_notification.html.erb:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
```
trailing slash isn't necessary when using HTML5 doctype.

JessSumner commented:
@halogenandtoast Ok I will investigate the flash messages. I just removed turbolinks from my app and I'm still seeing the issue.

halogenandtoast commented:
Also as a suggestion I'd cleanup your branches on github. You should probably only have master and one working branch.

halogenandtoast commented:
Which browser are you seeing the alignment issue in?

JessSumner commented:
@halogenandtoast I am seeing the error in Chrome. I just checked Safari and it works perfectly there. 
Having checked in Safari I also noticed another oddity. In Chrome, when I am typing my responses on the form, the font changes for some words but not others. In Safari this doesn't happen. I was assuming that was a separate issue but maybe they are both a browser issue loading my CSS files?

JessSumner commented:
@halogenandtoast Yes I will clean up my branches.

JessSumner commented:
@halogenandtoast I just read a forum post about disabling the cache would help. I submitted the form with the developer tools open and disable cache checked. It redirect worked perfectly.
Whilst that seems to have fixed it this time, presumably the error could happen again. Is there a way I can prevent this?

JessSumner commented:
@halogenandtoast I've answered one of my questions. As soon as I close the developer tools, I see the bug again.
Also I am seeing the text font change even with the developer tools open, so I will treat that as a separate issue.

kenyonj commented on app/helpers/contacts_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module ContactsHelper
+end
```
Can this file be removed as it's not being used for anything?

halogenandtoast commented:
@JessSumner this problem sounds crazy it doesn't happen in my Chrome. Do you perhaps have some extension that is interfering or maybe you've accidentally changed chromes font size for localhost? Command + 0 will reset it. Otherwise I think it would have to be something I could see before I would be able to make any suggestions.

kenyonj commented on app/controllers/contact_messages_controller.rb:
```diff
@@ -0,0 +1,31 @@
+      :name,
+      :email,
+      :cell,
+      :message
```
I like to append a comma after each item in these lists and then alphabetize the list. For me, it feels a little more organized.

halogenandtoast replied:
@kenyonj you would! I hate alphabetizing with a passion. Can't we sort them by length instead?

kenyonj replied:
I find it easier to compare two lists if they are alphabetized. For instance I have to add this list into a test, if it's alphabetized and I add an item I know where I need to add it, then next time I am not wondering if both lists match.

kenyonj replied:
what's your reason for hating it @halogenandtoast ?

halogenandtoast replied:
I was kidding about the length, but my main issue with alphabetizing is that it creates an artificial constraint that leads to disagreement. Typical scenarios are that you forget to alphabetize and then have to deal with signal to noise issues in PRs. Another scenario is the alphabetizing often gives the illusion that it's faster to search large lists visually rather than using your editor. In the long run I end up not caring either way, but I never alphabetize.

kenyonj replied:
I try and do it from the beginning, but I do agree that it creates unnecessary and possibly confusing noise in a PR where you decide to alphabetize everything. The organization benefit for me outweighs the other stuff. It's more of a "cooking in a clean kitchen" thing to me I think.

halogenandtoast replied:
I think a closer analogy would be to alphabetize all of your ingredients before you start cooking :p

JessSumner replied:
@kenyonj @halogenandtoast Whilst I like the organized nature of alphabetizing, I have them mirroring the layout of the form. There was one point where I changed the form but didn't update the params in the controller. I was easily able to see which ones I had missed. 
I would like my herbs and spices alphabetized though!

kenyonj commented on app/views/new_contact_message_mailer/contact_message_notification.html.erb:
```diff
@@ -0,0 +1,21 @@
+    <h1 style="color:#363691">A visitor to Any Rat Rescue has sent you a message.</h1>
+    <div>
+      <table style="border-collapse: collapse;">
+        <% ContactMessage::FORM_FIELDS.each do |field| %>
```
I have an idea of what you wanted to do with that helper file now? Thoughts on moving the constant access over to the helper?

JessSumner replied:
@kenyonj Yes I think I like that idea.

JessSumner replied:
@kenyonj I misunderstood. I thought you meant to move the constant over to be defined in the helper rather than the model. 
I think I would prefer to remove the helper and have the logic here. It's not particularly logic heavy or cumbersome. I think having another file for calling the constant would be more work than is necessary.

kenyonj replied:
:+1: make sense

kenyonj commented on app/views/new_contact_message_mailer/contact_message_notification.text.erb:
```diff
@@ -0,0 +1,5 @@
+A visitor to Any Rat Rescue has sent you a message
+
+<% ContactMessage::FORM_FIELDS.each do |field| %>
```
this can point to the helper method as well

kenyonj commented on config/routes.rb:
```diff
@@ -5,6 +5,7 @@
   resources :adoption_applications, only: [:new, :create]
   resources :volunteer_applications, only: [:new, :create]
   resources :newsletters, only: [:index]
+  resources :contact_messages, only: [:new, :create]
```
I also try and keep the top level `resources` alphabetized and grouped, so I would have an alphabetized list of `resources` then another alphabetized list of single `resource` routes.

JessSumner replied:
@kenyonj Done! I do like that use of alphabetizing. :+1:

kenyonj commented on test/mailers/previews/new_contact_message_mailer_preview.rb:
```diff
@@ -0,0 +1,7 @@
+    contact_message = ContactMessage.first
+    NewContactMessageMailer.contact_message_notification(contact_message)
+  end
+end
```
Love previews! Nice job :+1: :100:

JessSumner committed:
Fixing the text issues on redirects

Realised that some of the overlaps were occuring at points where I had put links. I had added <em> tags around the content I wanted to display for the link. Removing these has corrected the issue for that section.

Adding in the property word-break:normal has fixed the error happening in other places.

JessSumner committed:
Corrections and organizations

Deleted the contacts helper file as it is empty.
Organized the routes resources to be alphabetical.
Added a semicolon at the end of property declaration.
Removed trailing slash from html preview.

JessSumner commented:
@halogenandtoast 
![screen shot 2016-02-10 at 10 11 43 pm](https://cloud.githubusercontent.com/assets/11034063/12968708/7b83bcda-d043-11e5-8166-a4c62ab27b6f.png)
I have improved the home page one slightly as I had <em> tags on my links which weren't necessary. They were causing some distortion I think.
Matt noticed that the text wasn't being contained in it's div and found a CSS property word-break: normal. This seemed to fix it but I realize now we had developer tools open.
However as you can see from the picture there is still some issue going on. 
I tried the Cmd 0 but that hasn't changed it. Or was that specifically for the text font size change? That's showing up occasionally but less frequently.

halogenandtoast commented:
Run your page through an HTML validator for example https://validator.w3.org/

halogenandtoast commented:
Probably need to do direct input.

JessSumner committed:
Removed word-break property

JessSumner commented:
@halogenandtoast That's a cool tool. Thank you for pointing that out.
Not sure whether good or bad but no errors!
![screen shot 2016-02-10 at 11 13 51 pm](https://cloud.githubusercontent.com/assets/11034063/12969273/27419832-d04c-11e5-919b-ce662dc0651f.png)

JessSumner commented:
@halogenandtoast @MattMSumner I have switched the base-font I am using back to the default that bitters uses. I wrote a very long message and saw no changes in font and when I submitted the form the redirected to page loaded perfectly. This bug is obviously something to do with the custom font I had imported.

JessSumner commented:
![screen shot 2016-02-11 at 10 11 34 am](https://cloud.githubusercontent.com/assets/11034063/12979905/e310910c-d0a7-11e5-9e60-161cec1c9d58.png)

JessSumner committed:
Removed the custom fonts

Why:
Bug was found that the text was not loading correctly on a redirect.

How:
Removed the custom fonts and reverted to the default fonts provided by bitters.
I have run through the sequence where the bug occured a couple of times and seen no occurance of the bug.

Side effects:
The resource cards with the logos instead of a map need tweaking as this new font is slightly larger and has pushed one of the logos slightly off the card. This will need to be addressed as it looks messy. Otherwise this change has not affected the layout of any other the other pages.

JessSumner committed:
Removed all reference to the imported fonts

The navigation bar had the font defined in it's css file.
All references to the fonts have been removed.
